### PR TITLE
Data table integration

### PR DESCRIPTION
**Pull request by pseudotensor**
Anything non-optimal here?

To try it, do:
cd build ; cmake .. -DUSE_CUDA=ON -DGPU_COMPUTE_VER="61"  ; make -j ; cd ../python-package ; python setup.py install ; cd .. ; python tests/benchmark/testdt.py --algorithm=gpu_hist

Should do both numpy and dt versions and one can compare the time for DMatrix1 (train) and DMatrix2 (test) matrices.  Should be very similar, so not performance loss compared to inputting column-major or row-major numpy.  You can change do_ccont = False to True to have numpy as column-major.

I'll remove debug stuff once I'm happy and you guys comment.

### CODE REVIEW

File: src/c_api/c_api.cc [hunk 1]
```diff
@@ -502,12 +507,278 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,
...
+  // restore omp state
+  omp_set_num_threads(nthread_orig);
 
   mat.info.num_nonzero = mat.row_data_.size();
   *out  = new std::shared_ptr<DMatrix>(DMatrix::Create(std::move(source)));
   API_END();
 }
 
+
+
```
Comment (RAMitchell) [inline src/c_api/c_api.cc:518]: I think there's 3x the code here that you need. Try a function like:
```
float get_dt_value(void*col_ptr, wchar_t*type, bst_uint row){
  switch(type){
    Case int32 if ... return nan else return ...
  }
}
```

I don't think there needs to be any other switch statements or even a class.

**Comment by pseudotensor** [comment]
The class is needed to avoid duplicate code in casting to correct type from string stype.  The switch used to be based upon string compare, but slow to do that for every element.

**Comment by pseudotensor** [comment]
So point is to cast and string compare only per column, not every element.  But I might compress missing and get value like you said, returning all to be float nan.  good one there :)

File: python-package/xgboost/core.py [hunk 2]
```diff
@@ -609,13 +712,16 @@ def slice(self, rindex):
...
         """
-        res = DMatrix(None, feature_names=self.feature_names)
-        res.handle = ctypes.c_void_p()
-        _check_call(_LIB.XGDMatrixSliceDMatrix(self.handle,
-                                               c_array(ctypes.c_int, rindex),
-                                               c_bst_ulong(len(rindex)),
-                                               ctypes.byref(res.handle)))
-        return res
+        if isinstance(self.data, dt.DataTable):
+            ValueError("slice not implemented for DataTable")
```
Comment (st-pasha) [inline python-package/xgboost/core.py]: datatable itself supports slicing; so you could just say `return DMatrix(self.data[rindex, :])`.

File: src/c_api/c_api.cc [hunk 3]
```diff
@@ -502,12 +507,278 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,
...
+        return true;
+    }
+    else if(d->datacoltype==5 && d->datacol_int2[i]==GETNA<int16_t>()){
+        return true;
+    }
+    else if(d->datacoltype==6 && d->datacol_int8[i]==GETNA<int64_t>()){
+        return true;
+    }
+    else return false;
+}
```
Comment (st-pasha) [inline src/c_api/c_api.cc]: This all looks rather inefficient... `get_dt_value` and `is_dt_missing` are called within a tight loop, having lots of branching there kills the performance.
As a first step, I'd recommend the following: in class `datacol_struct` remove all members "datacol_*" and "datacoltype". Instead, create a hierarchy of classes `datacol_bool .. datacol_double` each deriving from common `datacol_struct` and exposing virtual methods `is_missing(int) -> bool` and `get_value(int) -> float`. Current constructor of `datacol_struct` would then become a factory method for instantiating one of these subclasses.

File: src/c_api/dt.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+inline T        GETNA() { return NULL;  }
+template<> inline int8_t   GETNA() { return NA_I1; }
+template<> inline int16_t  GETNA() { return NA_I2; }
+template<> inline int32_t  GETNA() { return NA_I4; }
+template<> inline int64_t  GETNA() { return NA_I8; }
+template<> inline uint8_t  GETNA() { return NA_U1; }
+template<> inline uint16_t GETNA() { return NA_U2; }
+template<> inline uint32_t GETNA() { return NA_U4; }
+template<> inline float    GETNA() { return NA_F4; }
+template<> inline double   GETNA() { return NA_F8; }
```
Comment (st-pasha) [inline src/c_api/dt.h]: You don't need all this complexity, just 4 constants `NA_I1`, `NA_I2`, `NA_I4` and `NA_I8`.

**Comment by st-pasha** [comment]
I would not recommend compressing `is_missing` and `get_value` methods into one, because then you'd need to make an extra call `isnan(value)` in order to determine whether the value is there

File: src/c_api/c_api.cc (hunk 3 above)
Comment (RAMitchell) [reply src/c_api/c_api.cc]: I think switch statements and virtual functions are effectively equivalent AFAIK. Virtual functions incur a vtable function lookup that the cpu can use branch prediction for. https://stackoverflow.com/a/449967

File: python-package/xgboost/core.py (hunk 2 above)
Comment (pseudotensor) [reply python-package/xgboost/core.py]: Actually, I shouldn't have excluded datatable there, it's not related.   This doesn't involve slicing the original data table object, just the DMatrix itself that is already created.  So I'll simply remove.

Comment (pseudotensor) [reply python-package/xgboost/core.py]: Same with get_dump().  I'm not sure why I did those checks, just a mistake/confusion at the time about what they were doing.

File: src/c_api/dt.h (hunk 4 above)
Comment (pseudotensor) [reply src/c_api/dt.h]: Did you say Ux is not implemented yet?  I forget, but that would justify what you are saying.  And what about F4/F8, I need those, but you are saying I don't really use them.  True.

**Comment by pseudotensor** [comment]
@st-pasha I guess point with compressing is_missing and get_value is currently get_value is only ever used after is_missing.  So might as well return both the value and the bool in is_dt_missing() so have that available.  That way don't have to have 2 calls is_missing and get_value and only have one.  Also avoids extra branching (even if will predict well) and lookup in memory (even though will be in cache).

**Comment by pseudotensor** [comment]
Ok, so I merged missing and get_value into one, because both were only ever done together.  But I setup the switch so the value is only returned if not missing.  Doesn't change performance really.

File: src/c_api/c_api.cc (hunk 3 above)
Comment (pseudotensor) [reply src/c_api/c_api.cc]: @st-pasha @RAMitchell After merging the is_missing and get_value, there is only one switch and it should be similar to any factory method.  I'll still ask you sometime how to go about the factory method, as it seems to defeat the purpose and require being explicit about which type I have.  The whole point of the class and switches is to input a string and get the correct casted value.  If you can do that with a factory in an easy way, I'll ask how.

But it seems like it wouldn't save on code and would just be a different way to order things.

**Pull request merged**